### DISCUSSION TITLE
How to check for `developer` mode in `update` code?

### DISCUSSION THREAD

**Discussion by abitrolly** [Homebrew/discussions#3152] [Tap maintenance and Homebrew development] [answered]
I want to show in `brew up` if `developer` mode is enabled. 

The logic is contained in `Library/Homebrew/cmd/developer.rb` It could probably be refactored into a separate function, but I don't know what is the best way call it from another command.
```ruby
    env_vars = []
    env_vars << "HOMEBREW_DEVELOPER" if Homebrew::EnvConfig.developer?
    env_vars << "HOMEBREW_UPDATE_TO_TAG" if Homebrew::EnvConfig.update_to_tag?
    env_vars.map! do |var|
      "#{Tty.bold}#{var}#{Tty.reset}"
    end

    case args.named.first
    when nil, "state"
      if env_vars.any?
        puts "Developer mode is enabled because #{env_vars.to_sentence} #{"is".pluralize(env_vars.count)} set."
      elsif Homebrew::Settings.read("devcmdrun") == "true"
        puts "Developer mode is enabled."
      else
        puts "Developer mode is disabled."
      end
```

https://github.com/Homebrew/brew/blob/c65bb5d12981267fada2c5644a02bf814a639188/Library/Homebrew/cmd/developer.rb#L33-L48

**Comment by SMillerDev** [comment]
You could just make an alias that does `brew developer && brew up<date|grade>`

**Reply by abitrolly** [reply]
I want to change the line in `brew up`.
```
-Updated Homebrew from 231379a8b to c65bb5d12.
+Updated Homebrew from 231379a8b to c65bb5d12 (developer mode is on).
```

**Comment by Bo98** [comment, marked as answer]
If you're talking about making this change within brew for a PR or something, I would indeed suggest refactoring into a function maybe in its own file that's included in `global` so it's available everywhere via `Homebrew.developer_mode?` or something. Not sure the best way to handle the reason part of that.

If this is for something you are developing for a third party tap or something: I suggest `Utils.safe_popen_read(HOMEBREW_BREW_FILE, "developer").chomp` which is the way to invoke that command and retrieve the result.

**Reply by abitrolly** [reply]
I am not familiar with Ruby import mechanics or `brew` code layout. I guess it is the `global` - https://github.com/Homebrew/brew/blob/master/Library/Homebrew/global.rb 

API docs say that `Homebrew.developer` already exists https://rubydoc.brew.sh/Homebrew.html but when I try to access it in `cmd/update-report.rb`, I get the ```Error: undefined method `developer' for Homebrew:Module```

If I understood the [Ruby modules](http://ruby-for-beginners.rubymonstas.org/advanced/modules.html) right, the extension of `Homebrew` module in [Library/Homebrew/cmd/developer.rb](https://github.com/Homebrew/brew/blob/c65bb5d12981267fada2c5644a02bf814a639188/Library/Homebrew/cmd/developer.rb) should make the `.developer` method available automatically, but it doesn't.

**Reply by Bo98** [reply]
> API docs say that `Homebrew.developer` already exists https://rubydoc.brew.sh/Homebrew.html but when I try to access it in `cmd/update-report.rb`, I get the `` Error: undefined method `developer' for Homebrew:Module ``

The command that implements `brew developer` does technically live at `Homebrew.developer` yes, but that's only because it has to be somewhere. This is not proper API so should not really be appearing on that page - I'll adjust that for the future.

> If I understood the [Ruby modules](http://ruby-for-beginners.rubymonstas.org/advanced/modules.html) right, the extension of Homebrew module in [Library/Homebrew/cmd/developer.rb](https://github.com/Homebrew/brew/blob/c65bb5d12981267fada2c5644a02bf814a639188/Library/Homebrew/cmd/developer.rb) should make the .developer method available automatically, but it doesn't.

Yes but only if `require "cmd/developer"` is done - but you should never really be requiring command files. The proper solution is to refactor it elsewhere - particularly given that function just prints strings to stdout rather than properly return a value in Ruby.

**Reply by abitrolly** [reply]
> Yes but only if require "cmd/developer" is done - but you should never really be requiring command files.

Got it. Creating some `devmode.rb` and then `require "devmode"` from `global.rb` should insert new functions in `Homebrew` global module. By the way, I can't find any reference where "cmd/developer" or "developer" is imported at all. How `brew.rb` finds it?

**Reply by abitrolly** [reply]
This leads to another question. If there is no direct `require "cmd/developer"` in `global.rb`, but some dependency imported from `global` still requires `cmd/developer`, will the global `Homebrew` module still get developer methods injected? If yes. then how to track the chain who've inserted methods?

**Reply by SMillerDev** [reply]
I believe it just requires only the command you run. So if you run  `brew developer` it'll require `cmd/developer`
